### PR TITLE
Add file-backed Todo demo backend

### LINKED ISSUE #27: Backend: implement Todo demo API and file-backed storage
from: coordinator 📋

Parent: #22

## Objective
Implement the backend half of the Todo demo workload described in issue #22.

## Context
The repository currently contains orchestrator scaffolding, setup docs, and specialist agents, but no application code. The agreed MVP is a small Node-based Todo app with minimal dependencies and CRUD support for create/list/update/delete.

Recommended backend footprint:
- `package.json` at repo root with `start` and `test` scripts
- `demo/todo/server.js`
- `demo/todo/lib/todo-store.js`
- `demo/todo/data/todos.json`

## Scope
1. Introduce the minimal Node runtime entrypoint and scripts.
2. Build a small HTTP JSON API under `/api/todos`:
   - `GET /api/todos`
   - `POST /api/todos`
   - `PATCH /api/todos/:id`
   - `DELETE /api/todos/:id`
3. Implement file-backed persistence suitable for a single-process demo.
4. Keep API contracts explicit and easy for frontend/QA to consume.
5. Avoid new runtime dependencies unless clearly justified.

## Proposed Todo shape
```json
{
  "id": "string",
  "title": "string",
  "completed": false,
  "createdAt": "ISO timestamp",
  "updatedAt": "ISO timestamp"
}
```

## Acceptance criteria
- The backend can serve Todo CRUD operations locally.
- Storage logic is isolated enough for direct testing.
- Frontend work can reliably target the API contract.
- A PR is opened with implementation notes and any tradeoffs.

## Notes
- Prefer Node built-ins (`http`, `fs`, `path`, `crypto`, `node:test`, `assert`) over Express or a database.
- Coordinate file paths and API semantics with the frontend and QA issues that follow from #22.

### PR DESCRIPTION
**Pull request by Purna88836**
## Summary
- add a minimal Node backend entrypoint and root `start` / `test` scripts with no runtime dependencies
- implement `/api/todos` CRUD handlers on top of a file-backed `TodoStore`
- add `node:test` coverage for both the storage layer and the HTTP API, plus README notes for local usage

## API contract
- `GET /api/todos` -> `200 { "todos": Todo[] }`
- `POST /api/todos` with `{ "title": string }` -> `201 { "todo": Todo }` and `Location: /api/todos/:id`
- `PATCH /api/todos/:id` with `{ "title"?: string, "completed"?: boolean }` -> `200 { "todo": Todo }`
- `DELETE /api/todos/:id` -> `204`
- validation failures return `400 { "error": string }`, unknown ids return `404 { "error": "Todo not found." }`, and unsupported methods return `405` with an `Allow` header

## Storage details
Todos are persisted in `demo/todo/data/todos.json` as a JSON array. `TodoStore` serializes mutations through an in-process promise chain, which is enough for the single-process demo the issue asked for and keeps the persistence logic directly testable.

## Tradeoffs
- This is intentionally single-process and file-backed; it avoids a database and third-party frameworks, but it is not appropriate for concurrent multi-instance deployment.
- The session environment exposed here did not allow command execution, so I added the `npm test` entrypoint and Node test coverage but could not execute the suite from this run.

Closes #27

### CODE REVIEW

**Comment by Purna88836** [comment]
from: qa 🤖

Summary review for PR #29: Good direction overall—API shape is clear, file writes are serialized, and the added tests cover the main CRUD flow. I’m requesting changes for two correctness issues before merge: (1) the broad `TypeError -> 400` mapping leaks server-side storage corruption as a client error, and (2) malformed percent-encoded todo IDs currently return `500` instead of a client error. The current tests also do not cover those edge cases yet.

**Comment by Purna88836** [comment]
from: qa 🤖

Review finding (demo/todo/server.js:38): Catching every `TypeError` as a `400` will also misclassify server-side storage corruption as a client error. `TodoStore._readTodos()` / `normalizeStoredTodo()` throw `TypeError` when the backing JSON file has the wrong shape, so a damaged `todos.json` would make `GET /api/todos` return `400` even though the request is valid. Please narrow the 400 mapping to request-validation failures only, or use a dedicated validation error type.

**Comment by Purna88836** [comment]
@orchestrator
from: qa 🤖

@orchestrator
from: qa ✅

/fleet @backend Please fix the review findings on PR #29:
- Narrow the `TypeError` -> `400` handling in `demo/todo/server.js` so invalid persisted todo data is treated as a server error instead of a client validation error.
- Handle malformed percent-encoded todo IDs (`decodeURIComponent` / `URIError`) as a client error and add API coverage for that case.

**Comment by Purna88836** [comment]
from: qa 🤖

Review finding (demo/todo/server.js:26): `decodeURIComponent()` throws `URIError` for malformed percent-encoding. A request like `/api/todos/%E0%A4%A` currently falls into the generic `500` path instead of returning a client error. Please catch `URIError` here (or classify it with the other request-validation errors) and add an API test for the malformed-path case.

File: demo/todo/server.js [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+
+      if (pathname === '/api/todos') {
+        await handleCollectionRoute(request, response, store);
+        return;
+      }
+
+      const todoIdMatch = pathname.match(/^\/api\/todos\/([^/]+)$/);
+
+      if (todoIdMatch) {
+        const todoId = decodeURIComponent(todoIdMatch[1]);
```
Comment (Purna88836) [inline demo/todo/server.js:26]: `decodeURIComponent()` throws `URIError` for malformed percent-encoding. A request like `/api/todos/%E0%A4%A` currently falls through to the generic `500` path even though the client sent the bad identifier. Please catch/classify that as a client error and add an API test for the malformed-path case.

File: demo/todo/server.js [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+      }
+
+      sendJson(response, 404, { error: 'Route not found.' });
+    } catch (error) {
+      if (error instanceof SyntaxError) {
+        sendJson(response, 400, { error: 'Request body must be valid JSON.' });
+        return;
+      }
+
+      if (error instanceof TypeError) {
```
Comment (Purna88836) [inline demo/todo/server.js:38]: Catching every `TypeError` as a `400` will also misclassify server-side storage corruption as a client error. `TodoStore._readTodos()` / `normalizeStoredTodo()` throw `TypeError` when persisted data has the wrong shape, so a damaged `todos.json` would make `GET /api/todos` return `400` for a valid request. Please narrow the `400` mapping to request-validation failures only, or introduce a dedicated validation error type.

**Comment by Purna88836** [comment]
from: coordinator 📋

Inline review comments are now on the PR for the two blocking correctness issues the `qa` review surfaced in `demo/todo/server.js`: malformed percent-encoded IDs currently return `500`, and the broad `TypeError -> 400` mapping misclassifies corrupted persisted data as a client error. A backend handoff has already been posted; `auto_merge` is disabled, so this should stay open until those findings are addressed.
